language_server/src/lsp/feature_language_features.cpp

```diff
@@ -483,47 +544,144 @@ void feature_language_features::document_symbol(const nlohmann::json& id, const
     auto document_uri = extract_document_uri(params);
 
     const auto limit = 5000LL;
-    auto symbol_list = ws_mngr_.document_symbol(document_uri.c_str(), limit);
-
-    response_->respond(id, "", document_symbol_list_json(symbol_list));
+    ws_mngr_.document_symbol(
+        document_uri.c_str(), limit, make_response(id, response_, [this](document_symbol_list symbol_list) {
+            return document_symbol_list_json(symbol_list);
+        }));
 }
 
 void feature_language_features::opcode_suggestion(const nlohmann::json& id, const nlohmann::json& params)
 {
     auto document_uri = extract_document_uri(params);
 
-    auto suggestions = nlohmann::json::object();
-
     bool extended = false;
     if (auto e = params.find("extended"); e != params.end() && e->is_boolean())
         extended = e->get<bool>();
 
-    if (auto opcodes = params.find("opcodes"); opcodes != params.end() && opcodes->is_array())
+    auto opcodes = params.find("opcodes");
+    if (opcodes == params.end() || !opcodes->is_array())
+    {
+        response_->respond(id,
+            "",
+            nlohmann::json {
+                { "uri", document_uri },
+                { "suggestions", nlohmann::json::object() },
+            });
+        return;
+    }
+
+    class composite_response_t
+    {
+        nlohmann::json m_id;
+        response_provider* m_response;
+        std::string m_document_uri;
+
+        nlohmann::json m_suggestions = nlohmann::json::object();
+
+        size_t m_pending_responses = 1;
+        bool m_failed = false;
+
+        void send()
+        {
+            if (m_failed)
+                return;
+
+            m_response->respond(m_id,
+                "",
+                nlohmann::json {
+                    { "uri", std::move(m_document_uri) },
+                    { "suggestions", std::move(m_suggestions) },
+                });
+        }
+
+    public:
+        composite_response_t(nlohmann::json id, response_provider* response, std::string document_uri)
+            : m_id(std::move(id))
+            , m_response(response)
+            , m_document_uri(std::move(document_uri))
+        {}
+
+        bool valid() const { return !m_failed; }
+        void error(int ec, const char* error)
+        {
+            m_failed = true;
+            m_response->respond_error(m_id, "", ec, std::string(error), {});
+        }
+
+        void provide(std::optional<std::pair<std::string, nlohmann::json>> r)
+        {
+            if (m_failed)
+                return;
+
+            if (r.has_value())
+                m_suggestions[r->first] = std::move(r->second);
+
+            if (--m_pending_responses == 0)
+                send();
+        }
+
+        void requests_submitted()
+        {
+            if (--m_pending_responses == 0)
+                send();
```

Comment:
Can it somehow happen that `provide()` is called sends the data while decrementing the `m_pending_responses` to `0` and then `requests_submitted()` is called?

Maybe `assert` might be warranted here?

Reply:
That most certainly happens right now, that's why the counter starts at 1. If this is a threading concern - this class assumes single-threaded handling.